File: GameServer/GameServerApp/GameServerApp/Proto/CreateProto/GameLevelEnterRequestProto.cs
//===================================================
//作    者：xxx
//创建时间：2019-05-18 15:13:23
//备    注：
//===================================================
using System.Collections;
using System.Collections.Generic;
using System;

/// <summary> 客户端发送进去游戏关卡消息 </summary>
public struct GameLevelEnterRequestProto : IProto
{
    public ushort ProtoCode { get { return 10401; } }

    /// <summary> 游戏Id </summary>
    public int GameChapterId;
    /// <summary> 难度等级 </summary>
    public byte Grade;
    /// <summary> 章节关卡Id </summary>
    public int GameLevelId;

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(GameChapterId);
            ms.WriteByte(Grade);
            ms.WriteInt(GameLevelId);
            return ms.ToArray();
        }
    }

    public static GameLevelEnterRequestProto GetProto(byte[] buffer)
    {
        GameLevelEnterRequestProto proto = new GameLevelEnterRequestProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.GameChapterId = ms.ReadInt();
            proto.Grade = (byte)ms.ReadByte();
            proto.GameLevelId = ms.ReadInt();
        }
        return proto;
    }
}

File: MMORPG_2/Assets/Script/Data/Proto/CreateLuaProto/GameLevelEnterRequestProto.lua
--客户端发送进去游戏关卡消息
GameLevelEnterRequestProto = { ProtoCode = 10401, GameChapterId = 0, Grade = 0, GameLevelId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
GameLevelEnterRequestProto.__index = GameLevelEnterRequestProto;

function GameLevelEnterRequestProto.New()
    local self = { }; --初始化self
    setmetatable(self, GameLevelEnterRequestProto); --将self的元表设定为Class
    return self;
end


--发送协议
function GameLevelEnterRequestProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.GameChapterId);
    ms:WriteByte(proto.Grade);
    ms:WriteInt(proto.GameLevelId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function GameLevelEnterRequestProto.GetProto(buffer)

    local proto = GameLevelEnterRequestProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.GameChapterId = ms:ReadInt();
    proto.Grade = ms:ReadByte();
    proto.GameLevelId = ms:ReadInt();

    ms:Dispose();
    return proto;
end